File: template_manager/template.py
from .document_part import DocumentPart
from .inner_environment import InnerEnvironment
from .shorthand import Shorthand
from .wobject import Wobject
from .outer_environment import OuterEnvironment


class Template:
    def __init__(self, name, version_code, version_name, description, implicit_outer_environment,
                 document_parts=None, wobjects=None, outer_environments=None, inner_environments=None,
                 shorthands=None):
        self.name = name
        self.version_code = version_code
        self.version_name = version_name
        self.description = description
        self.implicit_outer_environment = implicit_outer_environment

        self.document_parts = [DocumentPart(**dp) for dp in (document_parts or [])]

        self.wobjects = [Wobject(**ob) for ob in (wobjects or [])]

        self.classic_outer_environments = [OuterEnvironment(**oe) for oe in outer_environments.get("classic", [])]
        self.fragile_outer_environments = [OuterEnvironment(**oe) for oe in outer_environments.get("fragile", [])]

        self.inner_environments = [InnerEnvironment(**ie) for ie in (inner_environments or [])]

        self.shorthand_hash = Shorthand(**shorthands.get("hash"), type="#")
        self.shorthand_at = Shorthand(**shorthands.get("at"), type="@")

File: wuff/template/Template.cpp
//
// Created by Michal Janecek on 27.01.2024.
//

#include "Template.h"

void Template::deserialize(const YAML::Node& node) {
    if (!node["name"] || !node["version_code"] || !node["version_name"] || !node["description"] || !node["implicit_outer_environment"]) {
        throw std::runtime_error("Template YAML node is missing one or more required fields.");
    }

    name = node["name"].as<std::string>();
    version_code = node["version_code"].as<std::string>();
    version_name = node["version_name"].as<std::string>();
    description = node["description"].as<std::string>();
    implicit_outer_environment = node["implicit_outer_environment"].as<std::string>();

    // Deserialize DocumentParts
    if (node["document_parts"]) {
        for (const auto& dpNode : node["document_parts"]) {
            auto dp = std::make_unique<DocumentPart>();
            dp->deserialize(dpNode);
            document_parts.push_back(std::move(dp));
        }
    }

    // Deserialize Wobjects
    if (node["wobjects"]) {
        for (const auto& woNode : node["wobjects"]) {
            auto wo = std::make_unique<Wobject>();
            wo->deserialize(woNode);
            wobjects.push_back(std::move(wo));
        }
    }

    // Deserialize Classic Outer Environments
    if (node["classic_outer_environments"]) {
        for (const auto& oeNode : node["classic_outer_environments"]) {
            auto oe = std::make_unique<OuterEnvironment>();
            oe->deserialize(oeNode);
            classic_outer_environments.push_back(std::move(oe));
        }
    }

    // Deserialize Fragile Outer Environments
    if (node["fragile_outer_environments"]) {
        for (const auto& oeNode : node["fragile_outer_environments"]) {
            auto oe = std::make_unique<OuterEnvironment>();
            oe->deserialize(oeNode);
            fragile_outer_environments.push_back(std::move(oe));
        }
    }

    // Deserialize Inner Environments
    if (node["inner_environments"]) {
        for (const auto& ieNode : node["inner_environments"]) {
            auto ie = std::make_unique<InnerEnvironment>();
            ie->deserialize(ieNode);
            inner_environments.push_back(std::move(ie));
        }
    }

    // Deserialize Shorthands
    if (node["shorthands"]["hash"]) {
        shorthand_hash = std::make_unique<Shorthand>();
        shorthand_hash->deserialize(node["shorthands"]["hash"]);
    }

    if (node["shorthands"]["at"]) {
        shorthand_at = std::make_unique<Shorthand>();
        shorthand_at->deserialize(node["shorthands"]["at"]);
    }
}